modules/calib3d/src/usac.hpp

```diff
@@ -0,0 +1,744 @@
+public:
+    static Ptr<HomographyMinimalSolver4ptsQR> create(const Mat &points_);
+};
+
+class HomographyMinimalSolver4ptsGEM : public MinimalSolver {
+public:
+    static Ptr<HomographyMinimalSolver4ptsGEM> create(const Mat &points_);
+};
+
+//////////////////////////////////////// NON MINIMAL SOLVER ///////////////////////////////////////
+class NonMinimalSolver : public Algorithm {
+public:
+    /*
+     * Estimate models from non minimal sample
+     * models.size() == output
+     */
+    virtual int estimate (const std::vector<int>& sample, int sample_size,
+                          std::vector<Mat>& models, const std::vector<double>& weights) = 0;
+
+    /*
+     * Get minimal sample size required for non-minimal estimation.
+     */
+    virtual int getMinimumRequiredSampleSize() const = 0;
+
+    /*
+     * Get maximum number of possible solutions.
+     */
+    virtual int getMaxNumberOfSolutions () const = 0;
+};
+
+class HomographyNonMinimalSolver : public NonMinimalSolver {
+public:
+    static Ptr<HomographyNonMinimalSolver> create(const Mat &points_);
+};
+
+
+//////////////////////////////////////////////////////////////////////////////////////
+//////////////////////////////////////// DEGENERACY //////////////////////////////////
+class Degeneracy : public Algorithm {
+private:
+    std::vector<Mat> recovered_models;
+public:
+    virtual ~Degeneracy() override = default;
+
+    /*
+     * Check if sample causes degenerate configurations.
+     * For example, test if points are collinear.
+     */
+    inline virtual bool isSampleGood (const std::vector<int> &sample) const {
+        return true;
+    }
+
+    /*
+     * Check if model satisfies constraints.
+     * For example, test if epipolar geometry satisfies oriented constraint.
+     */
+    inline virtual bool isModelValid (const Mat &model, const std::vector<int> &sample) const {
+        return true;
+    }
+
+    /*
+     * Check if model is degenerate.
+     * Firstly test if sample is degenerate. Then check model itself.
+     * For example, for fundamental matrix estimation checks if no more than 5 points lie on the dominant plane
+     * and then check if model satisfies oriented constraint.
+     * This method could be combination of two previous functions, although is more general.
+     */
+    virtual bool isModelDegenerate (const Mat &model, const std::vector<int> &sample) const {
+        return false;
+    }
+
+    /*
+     * Check if model satisfies rank constraint. For example, homography matrix must have rank 3; fundamental or
+     * essential matrix must have rank 2.
+     */
+    virtual bool satisfyRankConstraint (const Mat &model) const {
+        return true;
+    }
+
+    /*
+     * Recover rank constraint.
+     * Primarily for epipolar geometry estimation. If matrix is of rank 3 then do SVD to get rank 2.
+     * Return: true if recovered successfully, false - otherwise.
+     */
+    virtual bool recoverRank (Mat &model) const {
```

Comment:
no const ?

Reply:
It shouldn't be constant, it modifies the model inside the function. For example, for F or E matrix.